src/contracts/Qearn.h

```diff
@@ -0,0 +1,876 @@
+        uint64 AmountOfUnlocking;
+        uint64 AmountOfReward;
+        uint64 AmountOfburn;
+        uint64 RewardPercent;
+        sint64 divCalcu;
+        uint32 indexOfinvocator;
+        uint32 t;
+        uint32 count_Of_last_vacancy;
+        uint32 locked_weeks;
+        uint32 startIndex;
+        uint32 endIndex;
+        
+    };
+
+    PUBLIC_PROCEDURE_WITH_LOCALS(unlock)
+
+        if(input.Locked_Epoch > QEARN_MAX_EPOCHS || input.Amount < QEARN_MINIMUM_LOCKING_AMOUNT) 
+        {
+
+            output.returnCode = QEARN_INVALID_INPUT_LOCKED_EPOCH;               //   if user try to unlock with wrong locked epoch, it should be failed to unlock.
+            return ;
+
+        }
+
+        locals.indexOfinvocator = QEARN_MAX_LOCKS;
+        locals.startIndex = state.EpochIndex.get(input.Locked_Epoch).startIndex;
+        locals.endIndex = state.EpochIndex.get(input.Locked_Epoch).endIndex;
+
+        for(locals.t = locals.startIndex ; locals.t < locals.endIndex; locals.t++) 
+        {
+
+            if(state.Locker.get(locals.t).ID == qpi.invocator()) 
+            { 
+                if(state.Locker.get(locals.t)._Locked_Amount < input.Amount) 
+                {
+
+                    output.returnCode = QEARN_INVALID_INPUT_UNLOCK_AMOUNT;  //  if the amount to be wanted to unlock is more than locked amount, it should be failed to unlock
+                    return ;  
+
+                }
+                else 
+                {
+                    locals.indexOfinvocator = locals.t;
+                    break;
+                }
+            }
+
+        }
+
+        if(locals.indexOfinvocator == QEARN_MAX_LOCKS) 
+        {
+            
+            output.returnCode = QEARN_EMPTY_LOCKED;     //   if there is no any locked info in state.Locker array, it shows this address didn't lock at the epoch (input.Locked_Epoch)
+            return ;  
+        }
+
+        /* the rest amount after unlocking should be more than MINIMUM_LOCKING_AMOUNT */
+        if(state.Locker.get(locals.indexOfinvocator)._Locked_Amount - input.Amount < QEARN_MINIMUM_LOCKING_AMOUNT) 
+        {
+            locals.AmountOfUnlocking = state.Locker.get(locals.indexOfinvocator)._Locked_Amount;
+        }
+        else 
+        {
+            locals.AmountOfUnlocking = input.Amount;
+        }
+
+        locals.RewardPercent = QPI::div(state._CurrentRoundInfo.get(input.Locked_Epoch)._Epoch_Bonus_Amount * 10000000ULL, state._CurrentRoundInfo.get(input.Locked_Epoch)._Total_Locked_Amount);
+        locals.locked_weeks = qpi.epoch() - input.Locked_Epoch - 1;
+        locals.divCalcu = QPI::div(locals.RewardPercent * locals.AmountOfUnlocking , 100ULL);
+
+        if(qpi.epoch() - input.Locked_Epoch >= 0 && locals.locked_weeks <= 3) 
+        {
+            locals.AmountOfReward = QPI::div(locals.divCalcu * QEARN_EARLY_UNLOCKING_PERCENT_0_3, 10000000ULL);
+            locals.AmountOfburn = QPI::div(locals.divCalcu * QEARN_BURN_PERCENT_0_3 , 10000000ULL);
```

Comment:
This block does not run if qpi.epoch() == input.Locked_Epoch, because locals.locked_weeks is 4294967295 then.
1. Please move all the qubic calculation (locals.RewardPercent, locals.divCalcu, locals.AmountOfReward, locals.AmountOfburn) to below all those ifs and only assign the PERCENT constants to new local vars in the ifs in order to improve readability and reduce code duplication.
2. In order to fix the locked_weeks underflow issue, I suggest to remove the first if completely and set QEARN_EARLY_UNLOCKING_PERCENT_0_3 and QEARN_BURN_PERCENT_0_3 as the default values for the new local variables.
3. I suggest to rename locked_weeks to lockedEpochs, because weeks is misleading here in my opinion. You may have locked QUs on Thursday and unlock them 12 days (more than 1 week) later, you still have 0 locked epochs.
4. All ifs below `if(locals.locked_weeks >= 4 && locals.locked_weeks <= 7)` should be `else if`.

Reply:
Thank you a lot. I have just fixed all issues.